Feature/container (#492)
argonne-lcf/user-guides: docs/polaris/containers/containers.md

qsub -v CONTAINER=mpich-4_latest.sif job_submission.sh
```

--8<-- [start:commoncontainerdoc]

## Recipe-Based Container Building

As mentioned earlier, you can build Apptainer containers from recipe files. Instructions are available [here](https://apptainer.org/docs/user/1.2/build_a_container.html#building-containers-from-apptainer-definition-files). See [available containers](#available-containers) for more recipes. 

> > Note: You can also build custom recipes by bootstrapping from prebuilt images. For e.g the first two lines in a recipe to use our custom Tensorflow implementation would be `Bootstrap: oras` followed by `From: ghcr.io/argonne-lcf/tf2-mpich-nvidia-gpu:latest`

Reviewer: ```suggestion
> Note: You can also build custom recipes by bootstrapping from prebuilt images. For e.g the first two lines in a recipe to use our custom TensorFlow implementation would be `Bootstrap: oras` followed by `From: ghcr.io/argonne-lcf/tf2-mpich-nvidia-gpu:latest`
```

## Available containers

If you just want to know what containers are available, here you go. 

> * Examples for running mpich/MPI containers can be found [here](https://github.com/argonne-lcf/container-registry/tree/main/containers/mpi/Polaris)

Reviewer: ```suggestion
* Examples for running MPICH containers can be found [here](https://github.com/argonne-lcf/container-registry/tree/main/containers/mpi/Polaris)
```

**Permission Denied Error**: If you encounter permission errors during the build

* Check your quota and delete any unnecessary files. 

* Clean up Apptainer cache, `~/.apptainer/cache`, and set the Apptainer tmp and cache directories as below. If your home directory is full and if you are building your container on a compute node, then set the tmpdir and cachedir to local scratch 
```bash
export BASE_SCRATCH_DIR=/local/scratch/ # FOR POLARIS
#export BASE_SCRATCH_DIR=/raid/scratch/ # FOR SOPHIA
export APPTAINER_TMPDIR=$BASE_SCRATCH_DIR/apptainer-tmpdir
mkdir $APPTAINER_TMPDIR
export APPTAINER_CACHEDIR=$BASE_SCRATCH_DIR/apptainer-cachedir/
mkdir $APPTAINER_CACHEDIR
```

* Make sure you are not on a directory accessed with a symlink, i.e. check if `pwd` and `pwd -P` returns the same path.

* If any of the above doesn't work, try running the build in your home directory.

**Mapping to rank 0 on all nodes**: Ensure that the container's MPI aligns with the system MPI. For e.g. follow the additional steps outlined in the [container registry documentation for MPI on Polaris](https://github.com/argonne-lcf/container-registry/tree/main/containers/mpi/Polaris)

> **libmpi.so.40 not found**: This can happen if the container's application has an OpenMPI dependency which is not currently supported on Polaris. It can also spring up if the containers base environment is not a Debian-based architecture such as Ubuntu. Ensure the application has an MPICH implementation as well. Also try removing `.conda/`, `.cache/`, and `.local/` folders from your home directory and rebuilding the container.

Reviewer: ```suggestion
**libmpi.so.40 not found**: This can happen if the container's application has an OpenMPI dependency which is not currently supported on Polaris. It can also spring up if the container's base environment is not a Debian-based architecture such as Ubuntu. Ensure the application has an MPICH implementation as well. Also try removing `.conda/`, `.cache/`, and `.local/` folders from your home directory and rebuilding the container.
```

 

## Troubleshooting Common Issues

- **Permission Denied Error**: If you encounter permission errors during the build
	- Check your quota and delete any unnecessary files. 

> 	- Clean-up apptainer cache, `~/.apptainer/cache`, and set the apptainer tmp and cache directories as below. If your home directory is full and if you are building your container on a compute node, then set the tmpdir and cachedir to local scratch 

Reviewer: ```suggestion
	- Clean-up Apptainer cache, `~/.apptainer/cache`, and set the Apptainer temporary and cache directories as below. If your home directory is full and if you are building your container on a compute node, then set the temporary and cache directories to local scratch 
```

I would recommend using "Apptainer", capitalized whenever referring to the project/concept outside of a code block, as the project does in their own literature: https://github.com/apptainer/apptainer
I think it makes our docs look more polished

Reviewer: also earlier in this file: 
> This is important for system (Polaris - Cray) mpich to bind to containers mpich. Set the following environment variables

to 

> This is important for the system's Cray MPICH distribution to bind to the MPICH installation inside a container. Set the following environment variables
